Fix: Convert templates to JSX (fixes #248)

swashbuck opened the pull request:
Fixes #248 

[//]: # (Delete Fix, Update, New and/or Breaking sections as appropriate)
### Fix
* A sentence describing each fix

[//]: # (Mention any other dependencies)

joe-allen-89 commented on templates/hotgraphicPopup.jsx:
```diff
@@ -0,0 +1,102 @@
+  return (
+    <div className='hotgraphic-popup__inner'>
+
+      <templates.hotgraphicPopupToolbar {...props} />
```
I think this should be moved to below the content to allow screen readers to jump directly to content rather than the toolbar.

swashbuck replied:
Agreed. Done.

joe-allen-89 commented on templates/hotgraphicPopupToolbar.jsx:
```diff
@@ -0,0 +1,82 @@
+
+        <div
+          className="hotgraphic-popup__count"
+          dangerouslySetInnerHTML={{ __html: itemCount }}
```
Needs aria-hidden set to true as the new carousel behavior from wcag changes read out the numbering on the back and next button.

swashbuck replied:
Done.

joe-allen-89 commented on templates/hotgraphicPopupToolbar.jsx:
```diff
@@ -0,0 +1,82 @@
+      </button>
+
+      {a11yConfig._options._isPopupWrapFocusEnabled &&
+      <a className="a11y-focusguard a11y-ignore a11y-ignore-focus" tabIndex="0" role="presentation">&nbsp;</a>
```
When using Jaws (2023) the next button highlight is much bigger than expected, I think it's related to the "\&nbsp;" as soon as that's removed the highlight returns to the normal size. Not sure if there's a reason we have that in there?

swashbuck replied:
Done.

joe-allen-89 reviewed (commented):
A couple of changes/queries.

joe-allen-89 commented on templates/hotgraphicLayoutPins.jsx:
```diff
@@ -0,0 +1,92 @@
+    onPinClicked
+  } = props;
+
+  const itemAriaLabel = (index, pin, title) => {
```
Rather than having the itemAriaLabel outside of the item, could we add this inside the _items.map (as with the accordion), something like the below:  

const ariaLabel = `${_isVisited ? visited + '. ' : ''} ${_useNumberedPins ? (index + 1) + '.' : ''} ${compile(title)}. Item ${(index + 1)} of ${_items.length}`;

joe-allen-89 commented on templates/hotgraphicLayoutPins.jsx:
```diff
@@ -0,0 +1,92 @@
+import React from 'react';
+import { classes } from 'core/js/reactHelpers';
```
additional import for compile 

`import { classes, compile } from 'core/js/reactHelpers';`

joe-allen-89 commented on templates/hotgraphicLayoutPins.jsx:
```diff
@@ -0,0 +1,92 @@
+              style={{ top: _top + '%', left: _left + '%' }}
+            >
+
+              <span className="aria-label" dangerouslySetInnerHTML={ itemAriaLabel(index, _pin, title) } />
```
Change to `dangerouslySetInnerHTML={{ __html: compile(ariaLabel) }}`

(if you agree with moving aria label into the _items.map)

joe-allen-89 reviewed (commented):
A few suggested changes to move aria label into _items.map to match accordion.

joe-allen-89 commented on templates/hotgraphicLayoutTiles.jsx:
```diff
@@ -0,0 +1,41 @@
+import React from 'react';
```
Import Adapt 
`import Adapt from 'core/js/adapt';`

joe-allen-89 commented on templates/hotgraphicLayoutTiles.jsx:
```diff
@@ -0,0 +1,41 @@
+import React from 'react';
+import { classes } from 'core/js/reactHelpers';
```
Import compile 
`import { classes, compile } from 'core/js/reactHelpers';`

joe-allen-89 commented on templates/hotgraphicLayoutTiles.jsx:
```diff
@@ -0,0 +1,41 @@
+import { classes } from 'core/js/reactHelpers';
+
+export default function HotgraphicLayoutTiles(props) {
+
```
Add visited aria 
`const visited = Adapt.course.get('_globals')?._accessibility?._ariaLabels.visited;`

joe-allen-89 commented on templates/hotgraphicLayoutTiles.jsx:
```diff
@@ -0,0 +1,41 @@
+  return (
+    <div className="hotgraphic__tile-item-container" role="list">
+
+      {_items.map(({ _index, _graphic, _isVisited, title }) =>
```
add _pin 

`{_items.map(({ _index, _graphic, _isVisited, title, _pin }) => {`

joe-allen-89 commented on templates/hotgraphicLayoutTiles.jsx:
```diff
@@ -0,0 +1,41 @@
+
+      {_items.map(({ _index, _graphic, _isVisited, title }) =>
+        <div className="hotgraphic__tile-item" role="listitem" key={_index}>
+
```
Similar to accordion, add aria label in _items.map 

`const ariaLabel = ${_isVisited ? visited + '. ' : ''} ${_pin.alt ? _pin.alt : title}. Item ${(_index + 1)} of ${_items.length};`

joe-allen-89 commented on templates/hotgraphicLayoutTiles.jsx:
```diff
@@ -0,0 +1,41 @@
+            onClick={onPinClicked}
+          >
+
+            <span className="aria-label">{title}</span>
```
Update to 

`<span className="aria-label" dangerouslySetInnerHTML={{ __html: compile(ariaLabel) }} />`

joe-allen-89 reviewed (commented):
The same changes suggested in pins layout.

swashbuck commented:
@joe-allen-89 I've moved the aria label into _items.map. Split up each bit of data into a separate variable, so hopefully it's still easy to read. Let me know what you think, though.

joe-allen-89 commented:
> @joe-allen-89 I've moved the aria label into _items.map. Split up each bit of data into a separate variable, so hopefully it's still easy to read. Let me know what you think, though.

Don't mind them on one line myself but happy to be overruled on that. Other than that all looking good and working as expected.

swashbuck commented on templates/hotgraphicLayoutTiles.jsx:
```diff
@@ -0,0 +1,52 @@
+
+      {_items.map(({ _index, _graphic, _isVisited, title }) => {
+
+        const visited = _isVisited ? visitedLabel + '. ' : '';
+        const itemTitle = title + '. ';
+        const itemCount = compile(globals._components._hotgraphic.item, { itemNumber: _index + 1, totalItems: _items.length });
+        const ariaLabel = visited + itemTitle + itemCount;
```
@joe-allen-89 Yeah, my thought was that it's is easier to see that it's three separate "things" if they're split up. One long concatenated string is somewhat difficult to process in my mind. Happy to change if everyone feels differently.
